### PR TITLE
vsphere ipi: download ova and upload ova to vsphere

### PR DESCRIPTION
**Pull request by jcpowermac**
This PR uses previous work from closed pull requests.  Modified to support the change in direction for importing an OVA into vSphere.

Previous work @jstuever: This change caches a copy of the vmware ova specified in rhcos.json in
preperation to upload the OVA into a Template.

New work:
- Using terraform-provider-vsphere as an example and source
  add a resource `vsphereprivate_import_ova`. This requires
  adding source from both terraform-provider-vsphere project and
  govmomi since required structs and functions are not exported.
  Reuse ova import implementation adapting for use within a
  terraform resource.
- Modify existing terraform to use the above resource
- Modify vsphere tfvars

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -51,6 +51,7 @@ require (
 	github.com/hashicorp/terraform-plugin-sdk v1.7.0
 	github.com/hashicorp/vault v1.3.0 // indirect
 	github.com/hinshun/vt10x v0.0.0-20180809195222-d55458df857c // indirect
+	github.com/jcpowermac/terraform-provider-vsphereprivate v0.0.0-20200314174956-1c9da6dfe728 // indirect
```
Comment (jcpowermac) [inline go.mod]: I need to fix this

Comment (abhinavdahiya) [reply go.mod]: I think we can keep the entire code in the installer repo like the azureprivatedns

Comment (jcpowermac) [reply go.mod]: @abhinavdahiya yeah its all under plugins/vsphereprivate just need to figure out how to fix it.

File: pkg/terraform/exec/plugins/vsphereprivate/archive.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
+/*
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/archive.go]: can we move this to `internal` `vpshereprivate/internal/archive.go` ?

File: pkg/terraform/exec/plugins/vsphereprivate/archive.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Modified From:
+// https://github.com/vmware/govmomi/blob/master/govc/importx/archive.go
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/archive.go]: let's use specific release tag or commit.

File: pkg/terraform/exec/plugins/vsphereprivate/config.go [hunk 4]
```diff
@@ -0,0 +1,571 @@
+// Modified From:
+// https://github.com/terraform-providers/terraform-provider-vsphere/blob/master/vsphere/config.go
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/config.go]: this file is already available as public in the `terraform-provider-vsphere` so why copy?

File: pkg/terraform/exec/plugins/vsphereprivate/config.go [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+
+/* NOTES:
+ * - Removed required functions or sections of functions that
+ *   require terraform-provider-vsphere internal helpers.
+ * - Changed package name
+ * - For the ova import we will not use the pbm client or the rest client
+ * - go linting issues
+ */
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/config.go]: hmm.. let's keep `//`

Comment (abhinavdahiya) [reply pkg/terraform/exec/plugins/vsphereprivate/config.go]: > sections of functions that require terraform-provider-vsphere internal helpers.

we should be able to use terraform-providers-vsphere/vsphere/config.go which can use the internal functions?

File: pkg/terraform/exec/plugins/vsphereprivate/config.go [hunk 6]
```diff
@@ -0,0 +1,571 @@
...
+	"github.com/vmware/govmomi/vapi/rest"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/pbm"
+	"github.com/vmware/govmomi/session"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/debug"
+	"github.com/vmware/govmomi/vim25/soap"
+	"github.com/vmware/govmomi/vim25/types"
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/config.go]: i think we need to fix the formatting here.

File: pkg/terraform/exec/plugins/vsphereprivate/provider.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+const defaultAPITimeout = time.Minute * 5
+
+// Provider returns a terraform.ResourceProvider.
+func Provider() terraform.ResourceProvider {
+	vsphereProvider := vsphere.Provider()
+
+	vsphereProvider.(*schema.Provider).ResourcesMap = map[string]*schema.Resource{
+		"vsphereprivate_import_ova": resourceVSpherePrivateImportOva(),
+	}
+
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/provider.go]: also remove the data source map ?

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+
+	"github.com/pkg/errors"
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/nfc"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/ovf"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/soap"
+	"github.com/vmware/govmomi/vim25/types"
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: nit: formatting fix

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+			"folder": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The name of the folder to locate the virtual machine in.",
+			},
+
+			// Tagging
+			//vSphereTagAttributeKey: tagsSchema(),
+			// Custom Attributes
+			//customattribute.ConfigKey: customattribute.ConfigSchema(),
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: i don't think we should have commented code.

File: pkg/tfvars/vsphere/vsphere.go [hunk 10]
```diff
@@ -3,6 +3,9 @@ package vsphere
 import (
 	"encoding/json"
 
+	"github.com/pkg/errors"
+
+	"github.com/openshift/installer/pkg/tfvars/internal/cache"
```
Comment (abhinavdahiya) [inline pkg/tfvars/vsphere/vsphere.go]: nit: fix formatting

File: pkg/terraform/exec/plugins/vsphereprivate/config.go (hunk 4 above)
Comment (jcpowermac) [reply pkg/terraform/exec/plugins/vsphereprivate/config.go]: @abhinavdahiya we need `vimClient` from the `VSphereClient` struct.
https://github.com/terraform-providers/terraform-provider-vsphere/blob/master/vsphere/config.go#L35

If there is a better way please let me know because I certainly didn't want to add more here than we need.

File: go.sum [hunk 11]
```diff
@@ -1788,6 +1788,7 @@ github.com/openshift/hashicorp-terraform v0.12.20-openshift-2/go.mod h1:R/dUWEZV
 github.com/openshift/hashicorp-terraform-plugin-sdk v1.6.0-openshift h1:lfMsjeRYTWs2FmLo9ak0fqV7ekzpKUtV2JkLKdlx3Zk=
 github.com/openshift/hashicorp-terraform-plugin-sdk v1.6.0-openshift/go.mod h1:H5QLx/uhwfxBZ59Bc5SqT19M4i+fYt7LZjHTpbLZiAg=
 github.com/openshift/imagebuilder v1.1.0/go.mod h1:9aJRczxCH0mvT6XQ+5STAQaPWz7OsWcU5/mRkt8IWeo=
+github.com/openshift/installer v0.16.1/go.mod h1:VWGgpJgF8DGCKQjbccnigglhZnHtRLCZ6cxqkXN4Ck0=
```
Comment (abhinavdahiya) [inline go.sum]: hmm.. this is odd, why do we have the our own repo in the go.sum?

File: pkg/terraform/exec/plugins/vsphereprivate/config.go (hunk 4 above)
Comment (abhinavdahiya) [reply pkg/terraform/exec/plugins/vsphereprivate/config.go]:
```go
type VSphereClient struct {
	// The VIM/govmomi client.
	vimClient *govmomi.Client
}

type ConfigWrapper struct {
	config *vsphere.Config
}

func NewConfig(d *schema.ResourceData) (*ConfigWrapper, error) {
	config, err := vsphere.NewConfig(d)
	if err != nil {
		return nil, err
	}
	return &ConfigWrapper{config}
}

// vimURL returns a URL to pass to the VIM SOAP client.
func (cw *ConfigWrapper) vimURL() (*url.URL, error) {
	u, err := url.Parse("https://" + cw.config.VSphereServer + "/sdk")
	if err != nil {
		return nil, fmt.Errorf("error parse url: %s", err)
	}

	u.User = url.UserPassword(c.User, c.Password)

	return u, nil
}


// Client returns a new client for accessing VMWare vSphere.
func (cw *ConfigWrapper) Client() (*vsphere.VSphereClient, error) {
	client := new(VSphereClient)

	u, err := c.vimURL()
	if err != nil {
		return nil, fmt.Errorf("error generating SOAP endpoint url: %s", err)
	}

	err = c.config.EnableDebug()
	if err != nil {
		return nil, fmt.Errorf("error setting up client debug: %s", err)
	}

	// Set up the VIM/govmomi client connection, or load a previous session
	client.vimClient, err = c.config.SavedVimSessionOrNew(u)
	if err != nil {
		return nil, err
	}

	log.Printf("[DEBUG] VMWare vSphere Client configured for URL: %s", c.VSphereServer)

	return client, nil
}
```

File: data/data/vsphere/variables-vsphere.tf [hunk 12]
```diff
@@ -65,3 +65,8 @@ variable "vsphere_control_plane_num_cpus" {
 variable "vsphere_control_plane_cores_per_socket" {
   type = number
 }
+
+variable "ova_filepath" {
+  type = string
```
Comment (abhinavdahiya) [inline data/data/vsphere/variables-vsphere.tf]: needs a description.

File: pkg/rhcos/vmware.go [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+		return "", err
+	}
+
+	baseURL := base.ResolveReference(image).String()
+
+	// TODO: Get Uncompressed SHA256 into rhcos.json
+	// Attach sha256 checksum to the URL.  Always provide the
+	// uncompressed SHA256; the cache will take care of
+	// uncompressing before checksumming.
+	//baseURL += "?sha256=" + meta.Images.VMware.UncompressedSHA256
```
Comment (abhinavdahiya) [inline pkg/rhcos/vmware.go]: since OVA file is not uncompressed, the SHA (https://github.com/openshift/installer/pull/3299/files#diff-89719a4e5e27791c48d41ee4b359b7daR46) should just work. right?

File: pkg/terraform/exec/plugins/vsphereprivate/internal/archive.go [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Modified From:
+// https://github.com/vmware/govmomi/blob/e651475773eba5139c706613b3b035377d6c1c30/govc/importx/archive.go
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/internal/archive.go]: would be very nice if we can list high level changes from the source..

File: pkg/terraform/exec/plugins/vsphereprivate/internal/archive.go [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+
+package vsphereprivate
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/internal/archive.go]: package should be internal.. ?

File: pkg/terraform/exec/plugins/vsphereprivate/internal/archive.go [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+
+// ReadOvf method
+func (f *ArchiveFlag) ReadOvf(fpath string) ([]byte, error) {
+	r, _, err := f.Open(fpath)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+
+	return ioutil.ReadAll(r)
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/internal/archive.go]: hmm.. this is going to load up a lot data into memory.. how big is OVF file typically :P

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 17]
```diff
@@ -0,0 +1,374 @@
...
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Description:  "",
+				Required:     true,
+				ValidateFunc: validation.NoZeroValues,
+			},
+			"filename": {
+				Type:         schema.TypeString,
+				Description:  "",
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: missing description.

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 18]
```diff
@@ -0,0 +1,374 @@
...
+	return &schema.Resource{
+		Create:        resourceVSpherePrivateImportOvaCreate,
+		Read:          resourceVSpherePrivateImportOvaRead,
+		Update:        resourceVSpherePrivateImportOvaUpdate,
+		Delete:        resourceVSpherePrivateImportOvaDelete,
+		SchemaVersion: 1,
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Description:  "",
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: missing description..

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 19]
```diff
@@ -0,0 +1,374 @@
...
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The name of the folder to locate the virtual machine in.",
+			},
+		},
+	}
+}
+
+// ImportOvaParams contains the vCenter objects required to import a OVA into vSphere.
+type ImportOvaParams struct {
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: nit: why is this public?

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 20]
```diff
@@ -0,0 +1,374 @@
...
+	ResourcePool *object.ResourcePool
+	Datacenter   *object.Datacenter
+	Datastore    *object.Datastore
+	Network      *object.Network
+	Host         *object.HostSystem
+	Folder       *object.Folder
+}
+
+func findImportOvaParams(client *vim25.Client, datacenter, cluster, datastore, network string) (*ImportOvaParams, error) {
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: there is a mix of context.TODO and ctx being passed to various functions in this functions...?

also why not use the defaultAPITimeout to create a context with a deadline?

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 21]
```diff
@@ -0,0 +1,374 @@
...
+	if !foundNetwork {
+		return nil, errors.Errorf("failed to find a host in the cluster that contains the provided network")
+	}
+
+	return importOvaParams, nil
+}
+
+// Used govc/importx/ovf.go as an example to implement
+// resourceVspherePrivateImportOvaCreate and upload functions
+// See: https://github.com/vmware/govmomi/blob/master/govc/importx/ovf.go#L196-L324
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: should be referencing a commit or version tag.

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 22]
```diff
@@ -0,0 +1,374 @@
...
+}
+
+func resourceVSpherePrivateImportOvaCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] %s: Beginning import ova create", d.Get("filename").(string))
+
+	ctx := context.TODO()
+	client := meta.(*VSphereClient).vimClient.Client
+	ovaTapeArchive := &internal.TapeArchive{Path: d.Get("filename").(string)}
+	archive := &internal.ArchiveFlag{}
+	archive.Archive = ovaTapeArchive
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]:
```suggestion
	archive := &internal.ArchiveFlag{Archive: &internal.TapeArchive{Path: d.Get("filename").(string)}}
```

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 23]
```diff
@@ -0,0 +1,374 @@
...
+
+	importOvaParams, err := findImportOvaParams(client,
+		d.Get("datacenter").(string),
+		d.Get("cluster").(string),
+		d.Get("datastore").(string),
+		d.Get("network").(string))
+	if err != nil {
+		return err
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: wrap the error in useful context..?

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 24]
```diff
@@ -0,0 +1,374 @@
...
+	if vm == nil {
+		return fmt.Errorf("error VirtualMachine not found")
+	}
+
+	return nil
+}
+
+func resourceVSpherePrivateImportOvaUpdate(d *schema.ResourceData, meta interface{}) error {
+	return resourceVSpherePrivateImportOvaRead(d, meta)
+}
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: these functions seem odd.

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 25]
```diff
@@ -0,0 +1,374 @@
...
+func resourceVSpherePrivateImportOvaRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*VSphereClient).vimClient.Client
+	moRef := types.ManagedObjectReference{
+		Value: d.Id(),
+		Type:  "VirtualMachine",
+	}
+
+	vm := object.NewVirtualMachine(client, moRef)
+	if vm == nil {
+		return fmt.Errorf("error VirtualMachine not found")
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: maybe include the `Value` that failed to be found.

File: pkg/tfvars/vsphere/vsphere.go [hunk 26]
```diff
@@ -3,7 +3,10 @@ package vsphere
 import (
 	"encoding/json"
 
+	"github.com/openshift/installer/pkg/tfvars/internal/cache"
+
 	vsphereapis "github.com/openshift/machine-api-operator/pkg/apis/vsphereprovider/v1alpha1"
+	"github.com/pkg/errors"
```
Comment (abhinavdahiya) [inline pkg/tfvars/vsphere/vsphere.go]: <std lib>

<external>

<installer repo>

File: pkg/tfvars/vsphere/vsphere.go [hunk 27]
```diff
@@ -48,7 +58,13 @@ func TFVars(sources TFVarsSources) ([]byte, error) {
...
+		OvaFilePath:       cachedImage,
+	}
+
+	cfgFormatted, err := json.MarshalIndent(cfg, "", "  ")
+	if err != nil {
+		return nil, err
 	}
 
-	return json.MarshalIndent(cfg, "", "  ")
+	return cfgFormatted, nil
```
Comment (abhinavdahiya) [inline pkg/tfvars/vsphere/vsphere.go]: why this change?

File: go.sum (hunk 11 above)
Comment (patrickdillon) [reply go.sum]: This suggests we might have a circular dependency. One of the other libraries we are importing here uses the installer. I don't see the culprit though so maybe I'm wrong.

File: pkg/terraform/exec/plugins/vsphereprivate/internal/archive.go (hunk 14 above)
Comment (jcpowermac) [reply pkg/terraform/exec/plugins/vsphereprivate/internal/archive.go]: @abhinavdahiya I reviewed what we were using archive for and we didn't need to copy it in - so I removed it.

File: pkg/terraform/exec/plugins/vsphereprivate/internal/archive.go (hunk 16 above)
Comment (jcpowermac) [reply pkg/terraform/exec/plugins/vsphereprivate/internal/archive.go]: @abhinavdahiya 

```
$ ls -alh ~/Downloads/rhcos-44.81.202003062006-0-vmware.x86_64.ova
-rw-rw-r--. 1 jcallen jcallen 840M Mar  6 16:59 /var/home/jcallen/Downloads/rhcos-44.81.202003062006-0-vmware.x86_64.ova
```

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go (hunk 24 above)
Comment (jcpowermac) [reply pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: I used the vsphere terraform provider as an example for the read though I greatly simplified from their version.
I wasn't quite sure if update was required, if not we can always remove it.

Comment (jcpowermac) [reply pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: Removed update and set to `ForceNew`.

**Comment by jcpowermac** [comment]
Tested with all the requested changes, everything looks good.

@patrickdillon and @abhinavdahiya when you have a moment if there is anything else please let me know. Thanks!

**Comment by abhinavdahiya** [comment]
The current commit split is little odd to me, there are additions that the next commit removes.
if you want to give credit, i think you can add co-auther in the commit message.

personally, i would like one commit or 3 (vpshereprivate, tfvars update, ova download and pass thorugh)

File: data/data/vsphere/main.tf [hunk 28]
```diff
@@ -25,10 +32,20 @@ data "vsphere_network" "network" {
 }
 
 data "vsphere_virtual_machine" "template" {
-  name          = var.vsphere_template
+  name          = vsphereprivate_import_ova.import.name
   datacenter_id = data.vsphere_datacenter.datacenter.id
 }
 
+resource "vsphereprivate_import_ova" "import" {
+  name       = "${var.cluster_id}-rhcos"
```
Comment (abhinavdahiya) [inline data/data/vsphere/main.tf]: do the machine objects need to match up with this name? If they do maybe add that to the commit message.

File: data/data/vsphere/variables-vsphere.tf [hunk 29]
```diff
@@ -65,3 +65,9 @@ variable "vsphere_control_plane_num_cpus" {
 variable "vsphere_control_plane_cores_per_socket" {
   type = number
 }
+
+variable "ova_filepath" {
```
Comment (abhinavdahiya) [inline data/data/vsphere/variables-vsphere.tf]: the rest of the variables are `vsphere_control_plane_*`

File: pkg/rhcos/vmware.go [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+	if err != nil {
+		return "", err
+	}
+
+	baseURL := base.ResolveReference(image).String()
+
+	// TODO: Get Uncompressed SHA256 into rhcos.json
+	// Attach sha256 checksum to the URL.  Always provide the
+	// uncompressed SHA256; the cache will take care of
+	// uncompressing before checksumming.
```
Comment (abhinavdahiya) [inline pkg/rhcos/vmware.go]: is this TODO actually required?

Comment (abhinavdahiya) [reply pkg/rhcos/vmware.go]: if yes, let's track that as github issue or JIRA issue instead of the TODO here.

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 31]
```diff
@@ -0,0 +1,384 @@
...
+	// generate the path e.g. /datacenter/host/cluster
+	clusterComputeResource, err := finder.ClusterComputeResource(ctx, clusterPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the network properties that is defined in ClusterComputeResource
+	// We need to know if the network name provided exists in the cluster that was
+	// also provided.
+	err = clusterComputeResource.Properties(context.TODO(), clusterComputeResource.Reference(), []string{"network"}, &ccrMo)
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: why the context.TODO ?

File: pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go [hunk 32]
```diff
@@ -0,0 +1,384 @@
...
+	return lease.Upload(ctx, item, f, opts)
+}
+
+func resourceVSpherePrivateImportOvaCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] %s: Beginning import ova create", d.Get("filename").(string))
+
+	// Since this is the same context that will be used for upload
+	// it should only set to TODO since we do not know how long
+	// it may take to upload the RHCOS ova to vSphere.
+	ctx := context.TODO()
```
Comment (abhinavdahiya) [inline pkg/terraform/exec/plugins/vsphereprivate/resource_vsphereprivate_import_ova.go]: shouldn't we have all requests use 5 min timeout context and the upload one use a separate TODO/Background context..

**Comment by abhinavdahiya** [comment]
> Tested with all the requested changes, everything looks good.
> 
> @patrickdillon and @abhinavdahiya when you have a moment if there is anything else please let me know. Thanks!

minor nits, otherwise LGTM

File: data/data/vsphere/main.tf [hunk 33]
```diff
@@ -25,10 +32,20 @@ data "vsphere_network" "network" {
 }
 
 data "vsphere_virtual_machine" "template" {
-  name          = var.vsphere_template
+  name          = vsphereprivate_import_ova.import.name
```
Comment (abhinavdahiya) [inline data/data/vsphere/main.tf]: what's with the `var.vsphere_template` now? is this no longer required so maybe we drop it.

Comment (jcpowermac) [reply data/data/vsphere/main.tf]: Thanks @abhinavdahiya caught an issue that I missed.  Yeah we still need vsphere_template since that holds what the name of the template will be.  I shouldn't have set the name of the import_ova and instead should have used the variable.  That has been fixed.

File: vendor/modules.txt [hunk 34]
```diff
@@ -919,6 +919,10 @@ github.com/keybase/go-crypto/rsa
 github.com/konsorten/go-windows-terminal-sequences
 # github.com/kr/fs v0.1.0
 github.com/kr/fs
+# github.com/kr/pretty v0.1.0
+github.com/kr/pretty
+# github.com/kr/text v0.1.0
+github.com/kr/text
```
Comment (patrickdillon) [inline vendor/modules.txt]: Not sure if this is anything to be concerned about. I'm guessing `kr` is a transitive dependence of `govmomi`?

Comment (jcpowermac) [reply vendor/modules.txt]: looks like it
```
Gopkg.lock:  name = "github.com/kr/pretty"
Gopkg.lock:    "github.com/kr/pretty",
Gopkg.toml:  name = "github.com/kr/pretty"
go.mod:replace github.com/kr/pretty v0.1.0 => github.com/dougm/pretty v0.0.0-20171025230240-2ee9d7453c02
go.mod: github.com/kr/pretty v0.1.0
govc/flags/output.go:   "github.com/kr/pretty"
vendor/github.com/kr/pretty/Readme:    import "github.com/kr/pretty"
vendor/github.com/kr/pretty/Readme:    http://godoc.org/github.com/kr/pretty
vendor/modules.txt:# github.com/kr/pretty v0.1.0 => github.com/dougm/pretty v0.0.0-20171025230240-2ee9d7453c02
vendor/modules.txt:github.com/kr/pretty
```

**Comment by jcpowermac** [comment]
Need to change the context

```
ERROR                                              
ERROR Error: failed to lease complete: Post https://vcsa.vmware.devcluster.openshift.com/sdk: context deadline exceeded 
ERROR                                              
ERROR   on ../../../../../../tmp/openshift-install-183900479/main.tf line 39, in resource "vsphereprivate_import_ova" "import": 
ERROR   39: resource "vsphereprivate_import_ova" "import" { 
ERROR                                              
```

**Comment by jcpowermac** [comment]
> Need to change the context
> 
> ```
> ERROR                                              
> ERROR Error: failed to lease complete: Post https://vcsa.vmware.devcluster.openshift.com/sdk: context deadline exceeded 
> ERROR                                              
> ERROR   on ../../../../../../tmp/openshift-install-183900479/main.tf line 39, in resource "vsphereprivate_import_ova" "import": 
> ERROR   39: resource "vsphereprivate_import_ova" "import" { 
> ERROR                                              
> ```

Fixed

File: pkg/terraform/exec/plugins/vsphereprivate/config.go [hunk 35]
```diff
@@ -0,0 +1,67 @@
...
+	vimClient *govmomi.Client
+}
+
+// ConfigWrapper - wrapping the terraform-provider-vsphere Config struct
+type ConfigWrapper struct {
+	config *vsphere.Config
+}
+
+// NewConfig function
+func NewConfig(d *schema.ResourceData) (*ConfigWrapper, error) {
```
Comment (patrickdillon) [inline pkg/terraform/exec/plugins/vsphereprivate/config.go]: I **think** you can simplify this even more:

embed  *vsphere.Config in ConfigWrapper

create an interface:
```go
type VSphereConfig interface{
    vimURL() (*url.URL, error)
    Client() (*VSphereClient, error)
}
```

change `NewConfig` to return the interface `VSphereConfig`

then you don't have to reimplement vimURL() or Client()
(Client looks a little different than the underlying implementation, but I assume it is just a subset and you left out things you didn't need)

File: pkg/asset/machines/worker.go [hunk 36]
```diff
@@ -322,8 +322,9 @@ func (w *Worker) Generate(dependencies asset.Parents) error {
 			mpool.Set(ic.Platform.VSphere.DefaultMachinePlatform)
 			mpool.Set(pool.Platform.VSphere)
 			pool.Platform.VSphere = &mpool
+			templateName := clusterID.InfraID + "-rhcos"
 
-			sets, err := vsphere.MachineSets(clusterID.InfraID, ic, &pool, string(*rhcosImage), "worker", "worker-user-data")
+			sets, err := vsphere.MachineSets(clusterID.InfraID, ic, &pool, templateName, "worker", "worker-user-data")
```
Comment (patrickdillon) [inline pkg/asset/machines/worker.go]: this is good. I will follow up with a PR to rename parameters in machineset to indicate that this is the template.
